vm/devices/user_driver/src/lockmem.rs

```diff
@@ -30,38 +32,58 @@ struct Mapping {
 }
 
 impl Mapping {
-    fn new(len: usize) -> std::io::Result<Self> {
+    fn new(len: usize) -> anyhow::Result<Self> {
+        // Create a ramfs file to back the mapping. This is necessary to ensure
+        // the memory is not moved (which is possible for ordinary
+        // tmpfs/anonymous allocations, even when mlocked).
+        //
+        // FUTURE: investigate other mechanisms for this, since some
+        // enterprising kernel developer may change ramfs to use movable memory
+        // one day. Ideally, we'd use a proper IOMMU, but that's still not
+        // available in the paravisor environment.
+        let file = fs_err::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_TMPFILE)
+            .open("/ramfs/")
+            .context("failed to crate ramfs file")?;
+
+        file.set_len(len as u64)
+            .context("failed to set ramfs file length")?;
+
         // SAFETY: No file descriptor or address is being passed.
         // The result is being validated.
         let addr = unsafe {
             libc::mmap(
                 std::ptr::null_mut(),
                 len,
                 libc::PROT_READ | libc::PROT_WRITE,
-                libc::MAP_PRIVATE | libc::MAP_ANONYMOUS | libc::MAP_LOCKED,
-                -1,
+                libc::MAP_SHARED | libc::MAP_LOCKED,
```

Comment:
MAP_ANONYMOUS should zeroinit pages according to the documentation but MAP_SHARED should not. LockMem::new should zeroinit pages anyway, do we know if all cases are covered? (assuming the doc is correct about this).